pkg/checkpoint/protocol.go

```diff
@@ -125,278 +85,216 @@ func NewProtocol(
-	eventsOut.PushBack(protobufs.EpochProgressEvent(p.moduleConfig.Ord, source, tt.EpochNr(msg.Epoch)))
+	// the node is considered to be a straggler and is sent a stable checkpoint to catch uparams.
+	checkpointpbdsl.EpochProgress(m, moduleConfig.Ord, from, epoch)
 
 	// If checkpoint is already stable, ignore message.
-	if p.stable() {
-		return eventsOut
+	if state.Stable(p) {
+		return nil
 	}
 
 	// Check snapshot hash
-	if p.stateSnapshotHash == nil {
+	if state.StateSnapshotHash == nil {
 		// The message is received too early, put it aside
-		p.pendingMessages[source] = msg
-		return eventsOut
-	} else if !bytes.Equal(p.stateSnapshotHash, msg.SnapshotHash) {
+		state.PendingMessages[from] = &checkpointpbtypes.Checkpoint{
+			Epoch:        epoch,
+			Sn:           sn,
+			SnapshotHash: snapshotHash,
+			Signature:    signature,
+		}
+		return nil
+	} else if !bytes.Equal(state.StateSnapshotHash, snapshotHash) {
 		// Snapshot hash mismatch
-		p.Log(logging.LevelWarn, "Ignoring Checkpoint message. Mismatching state snapshot hash.", "source", source)
-		return eventsOut
+		p.Log(logging.LevelWarn, "Ignoring Checkpoint message. Mismatching state snapshot hash.", "from", from)
+		return nil
 	}
 
-	// TODO: Only accept messages from nodes in membership.
-	//       This might be more tricky than it seems, especially when the membership is not yet initialized.
-
 	// Ignore duplicate messages.
-	if _, ok := p.signatures[source]; ok {
-		return eventsOut
+	if _, ok := state.SigReceived[from]; ok {
+		return nil
 	}
-	p.signatures[source] = msg.Signature
+	state.SigReceived[from] = struct{}{}
 
 	// Verify signature of the sender.
-	sigData := serializeCheckpointForSig(p.epoch, p.seqNr, p.stateSnapshotHash)
-	eventsOut.PushBack(cryptopbevents.VerifySigs(
-		p.moduleConfig.Crypto,
-		[]*cryptopbtypes.SignedData{sigData},
-		[][]byte{msg.Signature},
-		protobufs.SigVerOrigin(p.moduleConfig.Self),
-		[]t.NodeID{source},
-	).Pb())
-
-	return eventsOut
-}
-
-func (p *Protocol) applyNodeSigsVerified(result *cryptopb.SigsVerified) (*events.EventList, error) {
-
-	// A checkpoint only has one signature and thus each slice of the result only contains one element.
-	sourceNode := t.NodeID(result.NodeIds[0])
-	err := result.Errors[0]
-
-	if !result.AllOk {
-		p.Log(logging.LevelWarn, "Ignoring Checkpoint message. Invalid signature.", "source", sourceNode, "error", err)
-		p.signatures[sourceNode] = nil
-		return events.EmptyList(), nil
-	}
-
-	// Note the reception of a valid Checkpoint message from node `source`.
-	p.confirmations[sourceNode] = struct{}{}
-
-	// If, after having applied this message, the checkpoint became stable, produce the necessary events.
-	if p.stable() {
-		return p.announceStable(), nil
-	}
-
-	return events.EmptyList(), nil
-}
+	sigData := serializeCheckpointForSig(p.Epoch, p.SeqNr, state.StateSnapshotHash)
+	cryptopbdsl.VerifySig(m,
+		moduleConfig.Crypto,
+		sigData,
+		state.Signatures[from],
```

Comment:
```suggestion
		signature,
```
The `state.Signatures[from]` value is still `nil` at this point.

Reply:
Done.